src/System.Threading/tests/ReaderWriterLockSlimTests.cs

```diff
@@ -448,5 +448,67 @@ public static void ReleaseReadersWhenWaitingWriterTimesOut()
                 writeWaiterThread.Join();
             }
         }
+
+        [Fact]
+        public static void DontReleaseWaitingReadersWhenThereAreWaitingWriters()
+        {
+            using(var rwls = new ReaderWriterLockSlim())
+            {
+                rwls.EnterUpgradeableReadLock();
+                rwls.EnterWriteLock();
+                // Typical order of execution: 0
+
+                // Add a waiting writer
+                var threads = new Thread[2];
+                using(var beforeEnterWriteLock = new ManualResetEvent(false))
+                {
+                    var thread =
+                        new Thread(() =>
+                        {
+                            beforeEnterWriteLock.Set();
+                            rwls.EnterWriteLock();
+                            // Typical order of execution: 3
+                            rwls.ExitWriteLock();
+                        });
+                    thread.IsBackground = true;
+                    thread.Start();
+                    threads[0] = thread;
+                    beforeEnterWriteLock.WaitOne();
+                }
+
+                // Add a waiting reader
+                using(var beforeEnterReadLock = new ManualResetEvent(false))
+                {
+                    var thread =
+                        new Thread(() =>
+                        {
+                            beforeEnterReadLock.Set();
+                            rwls.EnterReadLock();
+                            // Typical order of execution: 4
+                            rwls.ExitReadLock();
+                        });
+                    thread.IsBackground = true;
+                    thread.Start();
+                    threads[1] = thread;
+                    beforeEnterReadLock.WaitOne();
+                }
+
+                // Wait for the background threads to block waiting for their locks
+                Thread.Sleep(1000);
```

Comment:
What happens if there's a blip and this isn't long enough?  In general we're trying to avoid timeouts like this in tests, both because a) they may not be long enough, and b) they cause inner loop tests to take longer.  If we must have this, can you mark the test as [OuterLoop]?  Thanks.